### ISSUE TITLE
给ncnn导入android hardware buffer作为输入避免拷贝的最佳实践是什么呢？

### ISSUE THREAD

**Issue by Gao-JF** [Tencent/ncnn#5531]
## detail | 详细描述 | 詳細な説明
我看到ncnn中有可以导入hardwareBuffer的类和方法，但是一直找不到官方的demo和文档，只能找到非官方的实现https://blog.csdn.net/d408550969/article/details/135288193。这篇文章没有完整的代码，我尝试使用这些零散的代码去复现opengl ncnn零拷贝输入，但是失败了。我发现AHardwareBuffer内的数据是正常的，但导入到VkImageMat后的数据是花屏的。

请问我应该怎么使用android hardware buffer作为输入，避免拷贝呢？

**Comment by whyb** [comment]
可以参考这个实现： https://github.com/Tencent/ncnn/blob/master/src/allocator.h#L407

**Comment by Gao-JF** [comment]
> 可以参考这个实现： https://github.com/Tencent/ncnn/blob/master/src/allocator.h#L407

请问有实现的示例代码或者文档吗？

**Comment by securekim** [comment]
Hi — picking this up two years late, but reporting end-to-end results on a current Adreno device since this issue is still open and there is no working public example anywhere.

## TL;DR

The YCbCr camera → AHB → `VkAndroidHardwareBufferImageAllocator` → `ImportAndroidHardwareBufferPipeline` → `VkMat` path **works correctly** on a Galaxy S25 Ultra (Adreno 830, Vulkan 1.3.284), **but only if `libncnn.a` was compiled with `__ANDROID_API__ >= 26`**.

The pre-built `ncnn-yyyymmdd-android-vulkan` archives on the releases page are built with `-DANDROID_PLATFORM=android-21` (see [`build-android.cmd:20`](https://github.com/Tencent/ncnn/blob/master/build-android.cmd#L20)), which compiles out every AHB symbol from `libncnn.a`. Any caller that links the official prebuilt and tries to use `VkAndroidHardwareBufferImageAllocator` / `VkImageMat::from_android_hardware_buffer` / `ImportAndroidHardwareBufferPipeline` gets `ld.lld: undefined symbol: ...` and the entire AHB API is effectively unreachable.

This may be the original root cause behind the garbled output reports — the only way to get the AHB code to *link* with the official prebuilt is to define `__ANDROID_API__` locally in the caller, which then resolves to header declarations that point at functions that were never compiled into the archive.

## Reproducer environment

- **Device** — Samsung Galaxy S25 Ultra (SM-S938N)
- **GPU** — Adreno (TM) 830
- **Vulkan** — 1.3.284, driver 512.800.64
- **`gi.support_VK_ANDROID_external_memory_android_hardware_buffer()`** — returns `5` (spec version 5)
- **ncnn** — tag `20260113`, rebuilt locally with `-DANDROID_PLATFORM=android-26 -DNCNN_VULKAN=ON`
- **NDK** — r29 (`29.0.14206865`)
- **Camera format** — `AIMAGE_FORMAT_YUV_420_888`, AHB `fmt=0x23` (Y8Cb8Cr8_420), `usage=0x20103` (= `CPU_READ_OFTEN` + `GPU_SAMPLED_IMAGE` + `CAMERA_WRITE`)

## Symbol-level proof the official prebuilt is missing AHB

```
$ llvm-nm --demangle ncnn-20260113-android-vulkan/arm64-v8a/lib/libncnn.a \
    | grep -iE 'android_hardware_buffer|AHardwareBuffer|ImportAndroid'
(no output — 0 matches)

$ llvm-strings .../libncnn.a | grep android_hardware_buffer
layout (binding = 0) uniform sampler2D android_hardware_buffer_image;
vec3 rgb = texture(android_hardware_buffer_image, pos).rgb * 255.f;
(the convert_ycbcr SPIR-V is baked in — but the C++ wrappers that drive it are gone)
```

After rebuilding the same tag with `-DANDROID_PLATFORM=android-26`:

```
$ llvm-nm --demangle .../libncnn.a \
    | grep -iE 'android_hardware_buffer|AHardwareBuffer|ImportAndroid' | wc -l
18
```

## Working minimal example (per camera frame)

```cpp
// 1. Camera side: AImageReader MUST be created with GPU_SAMPLED_IMAGE in
//    the usage mask, otherwise the AHB it produces is not Vulkan-importable.
AImageReader_newWithUsage(w, h, AIMAGE_FORMAT_YUV_420_888,
    AHARDWAREBUFFER_USAGE_CPU_READ_OFTEN
        | AHARDWAREBUFFER_USAGE_GPU_SAMPLED_IMAGE,
    /*maxImages*/3, &reader);

// 2. Inside onImageAvailable, after AImage_getHardwareBuffer(image, &ahb):
ncnn::VkAndroidHardwareBufferImageAllocator alloc(vkdev, ahb);
ncnn::VkImageMat src = ncnn::VkImageMat::from_android_hardware_buffer(&alloc);

ncnn::ImportAndroidHardwareBufferPipeline pipe(vkdev);
pipe.create(&alloc, /*type_to=*/1 /*PIXEL_RGB*/,
            /*rotate_from=*/1 /*identity*/, w, h, opt);

ncnn::VkMat dst;
dst.create(w, h, 3, /*elemsize=*/4u, /*elempack=*/1, opt.blob_vkallocator);

ncnn::VkCompute cmd(vkdev);
cmd.record_import_android_hardware_buffer(&pipe, src, dst);
// In production, feed dst straight into ex.input(...) and skip the download.
ncnn::Mat host;
cmd.record_download(dst, host, opt);
cmd.submit_and_wait();

// host is now W x H x 3 float32 RGB. AImage_delete may now run.
// pipe and alloc must outlive every cmd.submit that references them.
```

I dumped the downloaded `VkMat` back to host via `record_download` for sanity-checking and the resulting RGB matched the camera's actual scene exactly — no chroma shift, no banding, no garbling. The YCbCr → RGB conversion in `convert_ycbcr.comp` works as intended once the C++ wrappers that drive it are actually compiled in.

## Per-frame cost (Adreno 830, 640x480 input, n = 25 cycles)

- `VkAndroidHardwareBufferImageAllocator` ctor + `VkImageMat::from_android_hardware_buffer` — **0.27 ms median**
- `ImportAndroidHardwareBufferPipeline::create` — **24 ms median** ⚠️
- `VkMat::create` (dst) — < 0.01 ms
- `record_import_android_hardware_buffer` + `record_download` — 0.4 ms
- `submit_and_wait` (GPU work) — 25 ms median, high variance under co-load with ncnn matting inference on the same compute queue
- **total** — ~45 ms

`pipe.create` dominates and recreating it per frame is slower than a plain `cv::Mat` → `from_pixels` upload would be. The pipeline depends only on the immutable sampler (= the `samplerYcbcrConversion` derived from the AHB's `externalFormat`) and the `(type_to, rotate_from, target_w, target_h)` specialisation constants, all of which are stable for a given camera session in practice — I logged `format=0x23` and identical dims across 30 consecutive AHBs from the same `AImageReader`. A 1-pipeline-per-session cache should bring the steady-state cost down to roughly the `record + submit` line.

## Suggestions for follow-up (happy to PR any subset)

1. **Documentation + example.** No `examples/` entry and no `docs/how-to-use-and-FAQ/` page demonstrating the AHB path exist today. I have a ~150-line working minimal example I can clean up as `examples/ahb_zero_copy_inference.cpp` + `docs/how-to-use-and-FAQ/use-ncnn-with-android-hardware-buffer.md`.
2. **Build guidance.** A one-line note next to [`build-android.cmd`](https://github.com/Tencent/ncnn/blob/master/build-android.cmd) explaining that AHB users must pass `-DANDROID_PLATFORM=android-26` or higher (and ideally publishing an `android-26+` variant of the prebuilt archive) would have saved everyone in this issue several days.
3. **Pipeline + sampler caching helper** (lower priority, possibly out of scope here): a thin `ImportAndroidHardwareBufferSession` that internalises pipeline reuse across AHBs from the same source, since the per-frame `pipe.create` cost is currently a footgun.

Happy to break (1) out as a small PR first if there's interest.

**Comment by securekim** [comment]
Quick follow-up — I measured the per-AHB-pointer caching I mentioned at the bottom of the previous comment, since the conclusion there ("recreating per frame is slower than `cv::Mat` → `from_pixels` upload") only holds without caching.

## Caching strategy

`AImageReader` cycles through a small fixed pool of `AHardwareBuffer` pointers (3 distinct values for `maxImages=3` on this device), so a plain `std::unordered_map<AHardwareBuffer*, entry>` keyed by the AHB pointer is sufficient — no `ncnn` API change needed, and no mutex because the camera callback is the only writer. The one detail is that `VkAndroidHardwareBufferImageAllocator`'s ctor does not take an AHB ref, so the cache also has to `AHardwareBuffer_acquire(ahb)` on first sight to keep the buffer alive past `AImage_delete`:

```cpp
auto it = g_cache.find(ahb);
if (it == g_cache.end()) {
    AHardwareBuffer_acquire(ahb);
    auto* alloc = new VkAndroidHardwareBufferImageAllocator(vkdev, ahb);
    auto  src   = VkImageMat::from_android_hardware_buffer(alloc);
    auto* pipe  = new ImportAndroidHardwareBufferPipeline(vkdev);
    pipe->create(alloc, /*type_to=*/1 /*RGB*/, /*rotate_from=*/1, w, h, opt);
    g_cache[ahb] = { alloc, pipe, std::move(src) };
}
auto& e = g_cache[ahb];
// per-frame work below uses e.pipe / e.src
```

## Numbers (Adreno 830, 640×480, n = 40 cycles)

- **cache MISS** (3 cycles — one per unique AHB in the pool)
  - `allocator` + `from_ahb` + `pipe.create` total: 5–22 ms
  - `record_import` + `record_download`: ~0.4 ms
  - `submit_and_wait`: 2–5 ms (idle GPU at startup)
  - **total: 8–38 ms**
- **cache HIT** (37 subsequent cycles — pool stabilises to 2 actively rotating AHBs in steady state)
  - setup lookup: **2–7 µs** (≈ 5000× reduction from the 24 ms median in the previous comment)
  - `record_import` + `record_download`: ~0.5 ms
  - `submit_and_wait`: 2–44 ms (high variance — GPU compute queue is shared with the ncnn matting graph running concurrently)
  - **total: ~10–15 ms median**

So with caching the per-frame cost drops from ~45 ms to ~10–15 ms, and what's left is dominated by `submit_and_wait` rather than pipeline setup. That last item is *not* an AHB-import problem — it's GPU queue contention with the matting net on the same compute queue, which I expect to fold away when the imported `VkMat` goes straight into `ex.input(...)` and the matting graph stays on-GPU end-to-end (working on that next).

Happy to include the cache helper inside the example I proposed in (1).
